frappe/public/js/frappe/ui/messages.js

```diff
@@ -122,8 +122,36 @@ frappe.msgprint = function(msg, title) {
 
 		// setup and bind an action to the primary button
 		if (data.primary_action) {
-			frappe.msg_dialog.set_primary_action(__(data.primary_action.label || "Done"),
+			if (data.primary_action.server_action && typeof data.primary_action.server_action === 'string') {
+				frappe.msg_dialog.set_primary_action(__(data.primary_action.label || "Done"),
+					() => {
+						frappe.call({
+							method: data.primary_action.server_action,
+							args: {
+								args: data.primary_action.args
+							}
+						});
+					});
+			}
+
+			if (data.primary_action.client_action && typeof data.primary_action.client_action === 'string') {
+				let parts = data.primary_action.client_action.split('.');
+				let obj = window;
+				for (let part of parts) {
+					obj = obj[part];
+				}
+				frappe.msg_dialog.set_primary_action(
+					__(data.primary_action.label || "Done"),
+					() => {
+						if (typeof obj === 'function') {
+							obj(data.primary_action.args);
+						}
+					}
+				);
+			} else {
+				frappe.msg_dialog.set_primary_action(__(data.primary_action.label || "Done"),
 				data.primary_action.action);
+			}
```

Comment:
Simplify:
```js
if (data.primary_action.server_action && typeof data.primary_action.server_action === 'string') {
	data.primary_action.action = () => {
		frappe.call({
			method: data.primary_action.server_action,
			args: {
				args: data.primary_action.args
			}
		});
	}
}

if (data.primary_action.client_action && typeof data.primary_action.client_action === 'string') {
	let parts = data.primary_action.client_action.split('.');
	let obj = window;
	for (let part of parts) {
		obj = obj[part];
	}
	data.primary_action.action = () => {
		if (typeof obj === 'function') {
			obj(data.primary_action.args);
		}
	}
}

frappe.msg_dialog.set_primary_action(
	__(data.primary_action.label || "Done"),
	data.primary_action.action
);

```

Reply:
client action will always override server action if both are sent?

Reply:
Yeah, you can add an `else` to the second `if` block, but both shouldn't be set anyway.

Reply:
Done